core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotReferenceCreationValidator.scala

```diff
@@ -0,0 +1,89 @@
+package org.broadinstitute.dsde.rawls.snapshot
+
+import bio.terra.datarepo.model.{SnapshotModel, CloudPlatform => SnapshotCloudPlatform}
+import org.broadinstitute.dsde.rawls.RawlsException
+import org.broadinstitute.dsde.rawls.model.WorkspaceCloudPlatform.WorkspaceCloudPlatform
+import org.broadinstitute.dsde.rawls.model.{Workspace, WorkspaceCloudPlatform}
+
+import scala.jdk.CollectionConverters.CollectionHasAsScala
+
+object SnapshotReferenceCreationValidator {
+  def constructor(workspaceContext: Workspace, snapshot: SnapshotModel): SnapshotReferenceCreationValidator =
+    new SnapshotReferenceCreationValidator(workspaceContext, snapshot)
+}
+
+// thrown to disallow creation of a snapshot reference on an unsupported platform
+class UnsupportedPlatformException(message: String) extends RawlsException(message)
+// thrown to disallow creation of a snapshot reference across cloud boundaries
+class PlatformBoundaryException(message: String) extends RawlsException(message)
+// thrown to disallow creation of a snapshot reference across a protected data boundary
+class ProtectedDataException(message: String) extends RawlsException(message)
+
+class SnapshotReferenceCreationValidator(val workspaceContext: Workspace, val snapshot: SnapshotModel) {
+
+  // TODO: get this information from a more authoritative source rather than relying on the hardcoded bucket prefix
+  private def isWorkspaceProtected: Boolean = workspaceContext.bucketName.startsWith("fc-secure")
+
+  private def isSnapshotProtected: Boolean = snapshot.getSource.asScala.exists(_.getDataset.isSecureMonitoringEnabled)
+
+  private def snapshotPlatform: SnapshotCloudPlatform = {
+    val snapshotSourceDatasets = snapshot.getSource.asScala
+    val datasetCloudPlatforms = snapshotSourceDatasets.map(_.getDataset.getCloudPlatform).filter(_ != null).toSet
+
+    // assume GCP if datasets don't come from a source explicitly marked with another platform
+    if (datasetCloudPlatforms.isEmpty) {
+      return SnapshotCloudPlatform.GCP
+    }
```

Comment:
I don't know if this is safe - we should check with the TDR team

Reply:
@nmalfroy I'm trying to determine the cloud platform of a given TDR snapshot here, to decide whether or not to allow it to be referenced by a workspace.

I'm getting this particular `snapshot` from a call to `dataRepoDAO.getSnapshot` from `SnapshotService` which calls this method.

I tried to dig into the TDR code to understand what assumptions I could make here but couldn't get to a definitive understanding of whether I can assume `cloudPlatform` is always set.  Can you advise on if I should treat `cloudPlatform` as non-nullable, and get rid of this assumption that _if it's not present then it must be GCP_?

Reply:
Discussed this in Slack and can confirm `cloudPlatform` should always be set and its absence would be considered a bug.  This should let me unroll a bunch of this extra complexity.

Reply:
See resulting `WrappedSnapshot` for the simplified goodness.